jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/DefaultAuthenticatorFactory.java

```diff
@@ -53,27 +55,74 @@ public class DefaultAuthenticatorFactory implements Authenticator.Factory
-            authenticator = new SslClientCertAuthenticator(sslContextFactories.iterator().next());
-        }
+            case Authenticator.BASIC_AUTH -> new BasicAuthenticator();
+            case Authenticator.DIGEST_AUTH -> new DigestAuthenticator();
+            case Authenticator.FORM_AUTH -> new FormAuthenticator();
+            case Authenticator.SPNEGO_AUTH -> new SPNEGOAuthenticator();
+            case Authenticator.NEGOTIATE_AUTH -> new SPNEGOAuthenticator(Authenticator.NEGOTIATE_AUTH);  // see Bug #377076
+            case Authenticator.MULTI_AUTH -> getMultiAuthenticator(server, context, configuration);
+            case Authenticator.CERT_AUTH, Authenticator.CERT_AUTH2 ->
+            {
+                Collection<SslContextFactory> sslContextFactories = server.getBeans(SslContextFactory.class);
+                if (sslContextFactories.size() != 1)
+                    throw new IllegalStateException("SslClientCertAuthenticator requires a single SslContextFactory instances.");
+                yield new SslClientCertAuthenticator(sslContextFactories.iterator().next());
+            }
+            default -> null;
+        };
+    }
 
-        return authenticator;
+    private Authenticator getMultiAuthenticator(Server server, Context context, Authenticator.Configuration configuration)
+    {
+        SecurityHandler securityHandler = SecurityHandler.getCurrentSecurityHandler();
+        if (securityHandler == null)
+            return null;
+
+        String auth = configuration.getAuthenticationType();
+        if (Authenticator.MULTI_AUTH.equalsIgnoreCase(auth))
+        {
+            MultiAuthenticator multiAuthenticator = new MultiAuthenticator();
+
+            String authenticatorConfig = configuration.getParameter("org.eclipse.jetty.security.multi.authenticators");
+            for (String config : StringUtil.csvSplit(authenticatorConfig))
+            {
+                String[] parts = config.split(":");
+                if (parts.length != 2)
+                    throw new IllegalArgumentException();
+
+                String authType = parts[0].trim();
+                String pathSpec = parts[1].trim();
+
+                Authenticator.Configuration.Wrapper authConfig = new Authenticator.Configuration.Wrapper(configuration)
+                {
+                    @Override
+                    public String getAuthenticationType()
+                    {
+                        return authType;
+                    }
+                };
+
+                Authenticator authenticator = null;
+                List<Authenticator.Factory> authenticatorFactories = securityHandler.getKnownAuthenticatorFactories();
+                for (Authenticator.Factory factory : authenticatorFactories)
+                {
+                    authenticator = factory.getAuthenticator(server, context, authConfig);
```

Comment:
Is this good enough? Should we make Authentication.Factory have a getType method that returns the authType to compare before trying to getAuthentication?

Actually, why are you calling the factory here? Can't you just call getAuthenticator?

I'm confused by this... need explanation

Reply:
This is the `DefaultAuthenticatorFactory` so this is used when we just have an `<auth-method>` configured in `web.xml`.

So this case we had auth-method of `MULTI`, and we read the init param of `org.eclipse.jetty.security.multi.authenticators` which contains a comma separated list of `<auth-method>:<path-spec>`.

And then we have to create those `Authenticator`s which are contained within the `MultiAuthenticator`.

So we could have something like this in `web.xml`:
```
  <login-config>
    <auth-method>MULTI</auth-method>
    <realm-name>myRealm</realm-name>
  </login-config>
  <context-param>
    <param-name>org.eclipse.jetty.security.multi.login_path</param-name>
    <param-value>/</param-value>
  </context-param>
  <context-param>
    <param-name>org.eclipse.jetty.security.multi.authenticators</param-name>
    <param-value>OPENID:/login/oidc,SIWE:/login/siwe</param-value>
  </context-param>
  <context-param>
    <param-name>org.eclipse.jetty.security.siwe.login_path</param-name>
    <param-value>/login.html</param-value>
  </context-param>
  <context-param>
    <param-name>org.eclipse.jetty.security.openid.error_page</param-name>
    <param-value>/error</param-value>
  </context-param>
```

Reply:
My point is that these authentication factories have not been called before with the wrong type, so are you sure they return null all the time?   

Also why do only some Authentication types have factories?

Reply:
So the way this works in `SecurityHandler` is we run through the known authenticator factories and use the first one which returns a non-null `Authenticator`.

The `DefaultAuthenticatorFactory` returns non null for authentication types which are implemented in `jetty-security`, but openid and siwe will have different authenticator factories because they are in separate modules, and these are discovered with the service loader.

Reply:
OK I guess.... feels funny that we know the type, but we are not checking that the factory we use is actually of that type.  Are you sure there is nothing that can be done to firm that up?

Reply:
An `AuthenticatorFactory` can support many authentication types, so we could add a `boolean isAuthTypeSupported(String authType)` method, but this seems like an unnecessary complication.